frontend/crd/src/app/milestones-page/milestones/milestone.service.ts

```diff
@@ -1,26 +1,46 @@
 import { Injectable } from '@angular/core';
-import {HttpClient} from "@angular/common/http";
-import {Milestone, MilestoneJSON} from "../../../domain/Milestone";
-import {map, Observable} from "rxjs";
+import { HttpClient } from "@angular/common/http";
+import { Milestone, MilestoneJSON } from "../../../domain/Milestone";
+import { concatMap, finalize, map, Observable, of, ReplaySubject } from "rxjs";
 import { Endpoints, constructBackendRequest } from 'src/app/util/http-helper';
 
 @Injectable({
   providedIn: 'root'
 })
 export class MilestoneService {
 
+  private milestoneCache$ = new ReplaySubject<any>();
+  private hasBeenRequested = false;
+
   constructor(
     private http: HttpClient,
   ) {
   }
 
-  // Get all data items
-  getMilestones(): Observable<Milestone[]> {
-    return this.http.get<Milestone[]>(constructBackendRequest(Endpoints.MILESTONES))
-      .pipe(map((data: any) => {
-        return data.map((milestoneData: MilestoneJSON) => {
-          return new Milestone(milestoneData)
-        })
-      }))
+  getMilestones(forceRefresh?: boolean): any {
+    // return last value (i.e. cache) from ReplaySubject or add data to it
+    if (!this.hasBeenRequested || forceRefresh) {
+      this.hasBeenRequested = true;
+      const milestoneData = this.http.get<Milestone[]>(constructBackendRequest(Endpoints.MILESTONES))
+        .pipe(map((data: any) => {
+          return data.map((milestoneData: MilestoneJSON) => {
+            return new Milestone(milestoneData)
+          })
+      }));
+
+      //makes sure the HTTP is done before adding data to cache
+      const processedObservable = milestoneData.pipe(
+          concatMap((processedData: any) => {
+            this.milestoneCache$.next(processedData);
+            return of(processedData);
+          })
+        );
+
+      return processedObservable;
```

Comment:
just return the milestoneCache instead of a new observable

Reply:
I tried just doing this.milestoneCache.next(this.http.get(whatever)) but then the function just returned this.milestoneCache before the http request has retrieved the data and it ends up just passing an empty cache every time